File: cpp/26_reciprocal_cycles.cpp
#include <iostream>
#include <map>

int get_cycle_length(int n){
    // store previously seen remainders and digit positions
    std::map<int, int> seen;
    int rem = 1, count = 0;
    while (rem != 0){
        rem = rem * 10 % n;
        if (seen.find(rem) != seen.end()) return count - seen[rem];
        seen[rem] = count++;
    }
    // reciprocal terminates, no cycle
    return 0;
}

int reciprocal_cycles(){
    int res = 0, max_len = 0;
    for (int i = 2; i < 1000; i++){
        int len = get_cycle_length(i);
        if (len > max_len){
            max_len = len;
            res = i;
        }
    }
    return res;
}

int main() {
    std::cout << reciprocal_cycles() << "\n";
}

File: c#/_26_ReciprocalCycles.cs
using System;
using System.Collections.Generic;

public class _26_ReciprocalCycles{
	public static int Euler26(){
		int res = 0, max = 0;
		for (int i = 2; i < 1000; i++){
			int cycle = GetCycleLength(i);
			if (cycle > max){
				max = cycle;
				res = i;
			}
		}
		return res;
	}
	
	public static int GetCycleLength(int n){
		int rem = 1, count = 0;
		Dictionary<int, int> seen = new Dictionary<int, int>();
		while (rem != 0){
			rem = (rem * 10) % n;
			if (seen.ContainsKey(rem)){
				return count - seen[rem];
			}
			seen.Add(rem, count++);
		}
		return 0;
	}
	
	public static void Main(){
		Console.WriteLine(Euler26());
	}
}
